# Kraken2 Pipeline

Welcome to the **Kraken2 Pipeline** repository! Here, you’ll find a comprehensive guide that outlines all the essential commands and steps for analyzing long-read FASTQ files, starting from the initial installation of Conda to calculating abundance and relative abundance.

### What You’ll Find Here

In this repository, I've meticulously documented the entire process I followed to complete this analysis, including useful tips and best practices that I discovered along the way. 

If you’re embarking on your first project involving long reads, you’ve come to the right place! This repository is designed to provide a clear roadmap through all the fundamental steps you need to follow. I've curated these instructions from the original repositories for **Kraken2**, **Bracken**, and **Pavian**, ensuring you have a reliable resource at your fingertips.

### Why This Repository?

- **Step-by-Step Instructions**: Easy-to-follow commands that guide you through the entire workflow.
- **Useful Tips**: Insights and advice that can enhance your project and streamline your analysis.
- **Comprehensive Coverage**: All the essential tools and processes needed for long-read analysis in one place.

Get started today and make your long-read analysis project a success!

Kraken is a taxonomic sequence classifier that assigns taxonomic labels to DNA sequences by analyzing k-mers (short DNA sequences) in the query sequence. These k-mers are compared against a database that links them to the lowest common ancestor (LCA) of genomes containing that specific k-mer. The initial version of Kraken used a large indexed list of k-mer/LCA pairs to provide rapid classification, but its high memory requirements posed challenges for users. To address this, Kraken 2 was developed with enhanced efficiency and reduced memory needs.

Kraken 2 introduces several key improvements over Kraken 1. Instead of storing and querying all k-mers, Kraken 2 uses only minimizers—a smaller, representative subset of k-mers, referred to as ℓ-mers. These ℓ-mers are stored in a compact hash table, which uses a probabilistic data structure to further reduce memory consumption. This structure may occasionally result in false positives (incorrect LCA assignments or missed queries), but the error rate is low (under 1%) and can be mitigated by setting confidence scoring thresholds.

Kraken 2's database design, which stores only a fraction of possible ℓ-mers from genomic libraries, achieves high efficiency without compromising much accuracy. This approach is similar to Kraken 1's "MiniKraken" databases, but with significantly less loss in sensitivity per read. Testing shows that even reduced Kraken 2 databases maintain comparable accuracy to full-sized ones, solving a major limitation of the earlier version.
## What the project does?
The project has codes referred from original Kraken2 github file from initial installation to final report generation. So this gives the user guidance in following the project and easy understanding.
## Why the project is useful?
Following this project we can build a pipeline for analysing taxonomical classification of given raw reads.
## How users can get started with the project?
Run these codes on the server because building database requires about 100GB of storage and much computational power.
### System pre-rerquisites:
  **Disk space**: requires about 100GB disk space.
  **Memory**: The default database size is 29 GB (as of Jan. 2018), and you will need slightly more than that in RAM if you want to build the default database.
  **Dependencies**: It mostly uses Linux commands. Scripts are written mostly in bash shell, and perl. Downloads of NCBI data are performed by wget and rsync.
    Kraken 2 will attempt to use the dustmasker or segmasker programs provided as part of NCBI's BLAST suite to mask low-complexity regions. 

## Installing Conda
You can skip this step if your system already has miniconda or any other conda installed.
Getting miniconda installer using wget

```bash
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
```
Making script of installer to executable mode
```bash
chmod +x Miniconda3-latest-Linux-x86_64.sh
```

Run the installer
```bash
./Miniconda3-latest-Linux-x86_64.sh
```

Initialize conda and activate
```bash
conda init
conda activate
```

## Creating new Conda environment
Creating a new environment in Conda and working within a specific environment is always preferred because it helps manage dependencies across different versions, allows you to work in an isolated environment, ensures reproducibility, and keeps the base environment clean.
```bash
conda create -n <environment_name>
```
To rename the environment(if required)
```bash
conda rename <old_name> <new_name>
```

Activating the environment
```bash
conda activate <environment_name>
```

To check list of environments
```bash
conda env list
```
to know which environment is active look for '*" symbol

To exit from environment
```bash
conda deactivate
```

## Installing Kraken2 in the environment
Ensure that you are in correct envrionment before downloading and installing.
Download Kraken 2 Source Code
Use wget or curl to download the Kraken 2 source code from the official Kraken 2 GitHub repository:
```bash
wget https://github.com/DerrickWood/kraken2/archive/v2.1.2.tar.gz
```
Extract the tarball
```bash
tar -zxvf v2.1.2.tar.gz
```
Navigate to the Extracted Directory
```bash
cd kraken2-2.1.2
```
Install Kraken 2 in the chosen directory
```bash
./install_kraken2.sh /path/to/install/kraken2
```
Add kraken2 path to our path
```bash
export PATH="/path/to/install/kraken2:$PATH"
```
Check whether kraken2 installed properly or not
```
kraken2 -v
```
