# Bracken Installation and Usage Guide

## Introduction

Bracken is a tool designed to improve taxonomic classification in metagenomics analyses by leveraging Kraken2 reports. While Kraken2 classifies reads, it sometimes generates ambiguous classifications, denoted by codes such as P1, P4, and C1. Bracken addresses this limitation by applying statistical methods to reassign these ambiguous reads to appropriate taxa based on the Kraken2 report.

### Prerequisites

Before installing Bracken, ensure that you have already installed Kraken2, as Bracken operates as an additional tool for it. Also, make sure you are working within your Conda environment. You can check your current Conda environment with:

```bash
conda env list
```
An asterisk (*) will indicate your active environment.

# Installation Steps
Now we will move on to installation.

Check if Bracken is installed anywhere by:
```bash
bracken -v
```
If the version number does not appear, you will need to install Bracken.

## Clone the Bracken Repository

Use Git to clone the Bracken repository from GitHub:
```bash
git clone https://github.com/jenniferlu717/Bracken.git
```
## Navigate to the Bracken Directory

Change your current directory to the newly cloned Bracken folder:

```bash
cd Bracken/
```
## Install Bracken
Run the installation script provided in the repository:

```bash
bash install_bracken.sh
```
## Update Your PATH
To make Bracken accessible from any terminal session, add the Bracken directory to your PATH. Replace /path/to/Bracken with the actual path where Bracken is installed:

```bash
export PATH="$PATH:/path/to/Bracken"
```
You can add this line to your shell configuration file (e.g., ~/.bashrc or ~/.zshrc) for persistence across sessions.

## To verify the installation
```bash
bracken -v
```
It should give you version name of Bracken.

## Build database using Kraken2 db

```bash
bracken-build -d <path to kraken2 database> -t <thread numbers> -k <35 by default> -l <100 or any length of read>
```
Parameters:
-d <path to kraken2 database>: Path to your Kraken2 database.
-t <number of threads>: Number of threads to use for processing.
-k <kmer length>: Kmer size; the default is 35.
-l <length of reads>: Specify the length of the reads (e.g., 100).

## Finding abundance
Finding Abundance
Once you have your Kraken2 report ready, you can calculate the abundance of taxa using Bracken with the following command:

```bash
bracken -d <path to kraken2 database> -i <path to Kraken2 report> -o <name of output file> -r 100 -l <level of taxon S/G/F/O/C/P/K/D> -t <threshhold number like 10>
```
Here,
-d <path to kraken2 database>: Specifies the path to the Kraken2 database that was used to generate the report. Bracken will use this database to calculate the abundance estimates.
-i <path to Kraken2 report>: Indicates the path to the Kraken2 report file (usually a .report file). This report contains the classification results that Bracken will analyze to determine the abundance of taxa.
-o <name of output file>: Specifies the name of the output file where Bracken will write the abundance results. This file will typically have a .tsv extension (Tab-Separated Values) and will contain the abundance data for the specified taxon level.
-r 100: Sets the read length (in base pairs) for which to calculate abundance. This value is used to determine how many reads are considered for estimating the abundance of taxa. The default value is often set to 100, but it can be adjusted based on your dataset.
-l <level of taxon S/G/F/O/C/P/K/D>: Indicates the taxonomic level at which to calculate abundance. The levels are:
S - Species
G - Genus
F - Family
O - Order
C - Class
P - Phylum
K - Kingdom
D - Domain
You would replace <level of taxon> with the appropriate letter corresponding to the taxonomic level you wish to analyze.

-t <threshold number>: Sets a threshold for the minimum number of reads needed to consider a taxon as present in the results. For example, if set to 10, only taxa with 10 or more reads will be included in the output.

## Copying Output Files
To copy or move the output file from the server to your local disk, open WSL on your local machine and navigate to your desired directory. Connect to the server using the following command:

```bash
scp <username>@<IP address>://<path to the bracken output file>/*.bracken ./
```

## Conclusion
After running the Bracken commands, you can examine the output file, which will include a column for "added reads." These reads represent the ambiguous classifications from the Kraken2 report that Bracken has successfully reclassified to the appropriate taxa.

Feel free to explore the Bracken documentation for further details on its usage and features. If you encounter any issues or have questions, don't hesitate to ask for help!
**https://github.com/jenniferlu717/Bracken/tree/master**
